extension/json/json_scan.cpp

```diff
@@ -0,0 +1,497 @@
+		json_readers.push_back(make_unique<BufferedJSONReader>(context, bind_data.options, i, bind_data.file_paths[i]));
+	}
+}
+
+unique_ptr<GlobalTableFunctionState> JSONScanGlobalState::Init(ClientContext &context, TableFunctionInitInput &input) {
+	auto &bind_data = (JSONScanData &)*input.bind_data;
+	return make_unique<JSONScanGlobalState>(context, bind_data);
+}
+
+idx_t JSONScanGlobalState::MaxThreads() const {
+	return system_threads;
+}
+
+JSONScanLocalState::JSONScanLocalState(ClientContext &context, JSONScanGlobalState &gstate)
+    : json_allocator(BufferAllocator::Get(context)), current_reader(nullptr), current_buffer_handle(nullptr),
+      buffer_size(0), buffer_offset(0), prev_buffer_remainder(0) {
+
+	// Buffer to reconstruct JSON objects when they cross a buffer boundary
+	reconstruct_buffer = gstate.allocator.Allocate(gstate.bind_data.maximum_object_size);
+
+	// This is needed for JSONFormat::UNSTRUCTURED, to make use of YYJSON_READ_INSITU
+	current_buffer_copy = gstate.allocator.Allocate(gstate.buffer_capacity);
+	buffer_copy_ptr = (const char *)current_buffer_copy.get();
+}
+
+unique_ptr<LocalTableFunctionState> JSONScanLocalState::Init(ExecutionContext &context, TableFunctionInitInput &input,
+                                                             GlobalTableFunctionState *global_state) {
+	auto &gstate = (JSONScanGlobalState &)*global_state;
+	return make_unique<JSONScanLocalState>(context.client, gstate);
+}
+
+static inline void SkipWhitespace(const char *buffer_ptr, idx_t &buffer_offset, idx_t &buffer_size) {
+	for (; buffer_offset != buffer_size; buffer_offset++) {
+		if (!std::isspace(buffer_ptr[buffer_offset])) {
+			break;
+		}
+	}
+}
+
+idx_t JSONScanLocalState::ReadNext(JSONScanGlobalState &gstate) {
+	json_allocator.Reset();
+
+	idx_t count = 0;
+	if (buffer_offset == buffer_size) {
+		bool first_read;
+		if (!ReadNextBuffer(gstate, first_read)) {
+			return 0;
+		}
+
+		if (!first_read && current_reader->GetOptions().format == JSONFormat::NEWLINE_DELIMITED) {
+			ReconstructFirstObject(gstate);
+			count++;
+		}
+	}
+
+	auto &options = current_reader->GetOptions();
+	switch (options.format) {
+	case JSONFormat::UNSTRUCTURED:
+		ReadUnstructured(count);
+		break;
+	case JSONFormat::NEWLINE_DELIMITED:
+		ReadNewlineDelimited(count, gstate.bind_data.ignore_errors);
+		break;
+	default:
+		throw InternalException("Unknown JSON format");
+	}
+
+	// Skip over any remaining whitespace for the next scan
+	SkipWhitespace(buffer_ptr, buffer_offset, buffer_size);
+
+	return count;
+}
+
+static inline const char *NextNewline(const char *ptr, idx_t size) {
+	return (const char *)memchr(ptr, '\n', size);
+}
+
+static inline const char *PreviousNewline(const char *ptr) {
+	for (ptr--; true; ptr--) {
+		const auto &c = *ptr;
+		if (c == '\n') {
+			break;
+		}
+	}
+	return ptr;
+}
+
+static inline void TrimWhitespace(JSONLine &line) {
+	while (line.size != 0 && std::isspace(line[0])) {
```

Comment:
Can we use `StringUtil::CharacterIsSpace` here? `std::isspace`is [locale sensitive](https://en.cppreference.com/w/cpp/string/byte/isspace) which is evil.